cardano-mempool/src/Cardano/Memory/Pool.hs

```diff
@@ -0,0 +1,217 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+module Cardano.Memory.Pool where
+
+import Control.Monad
+import Control.Monad.Primitive
+import Foreign.Ptr
+import Foreign.ForeignPtr
+import GHC.ForeignPtr
+import Control.Applicative
+import Data.Bits
+import Data.Maybe
+import GHC.TypeLits
+import Data.Primitive.PrimArray
+import Data.Primitive.PVar
+import Data.Primitive.PVar.Unsafe (atomicModifyIntArray#)
+import Data.IORef
+import GHC.Int
+import GHC.IO
+import GHC.Exts (fetchAndIntArray#)
+
+data Block (n :: Nat) = Block
+
+blockByteCount :: KnownNat n => Block n -> Int
+blockByteCount = fromInteger . natVal
+
+data Page n =
+  Page
+    { pageMemory :: !(ForeignPtr (Block n))
+    , pageBitArray :: !(MutablePrimArray RealWorld Int) -- Need to use Int, since there are
+                                                        -- no built-in atomic ops for Word
+    , pageFull :: !(PVar Int RealWorld)
+    , pageNextPage :: !(IORef (Maybe (Page n)))
+    }
+
+data Pool n =
+  Pool
+    { poolFirstPage :: !(Page n)
+    , poolPageInitializer :: !(IO (Page n))
+    , poolBlockFinalizer :: !(Ptr (Block n) -> IO ())
+    }
+
+ixBitSize :: Int
+ixBitSize = finiteBitSize (0 :: Word)
+
+initPool ::
+     forall n. KnownNat n
+  => Int
+  -- ^ Number of groups per page. Must be a posititve number, otherwise error. One group
+  -- contains as many blocks as the operating system has bits. A 64bit architecture will
+  -- have 64 blocks per group. For example, if program is compiled on a 64 bit OS and you
+  -- know ahead of time the maximum number of blocks that will be allocated through out
+  -- the program, then the optimal value for this argument will @maxBlockNum/64@
```

Comment:
I think it would be better to have the caller pass in a page size, rather than "groups per page" here.

"Groups", as a concept, are an implementation detail, meaningless to the API consumer. What we care about on the outside is block size and page size (i.e., allocation granularity), and it's kind of clumsy to have to reverse the math to get the correct number of groups per page to pass to this function.

Reply:
Unfortunately page size is a not a good idea, because if the value supplied is not divisible by block size exactly we will end up allocating extra memory that will never be used.

Using group size gives the optimal utilization of the page, because it ensures that the page size is divisible exactly by the (number of blocks * width of atomic element in the bit array). So we utilize the page and the bit array more effectively.

That being said, we could switch to block count, rather than the group size. This will require masking off the last `Int` in the bit array, but it is doable. It would also provide a finer granularity of the page allocator. Let me know if you'd like me to switch to block count.

Reply:
Ah, I see where the confusion stems from.

In our use case, any leftover memory in an OS page that we don't use is going to be "wasted" anyway, because mlock() can only operate at OS page granularity, and our limiting factor is not overall memory use, but mlocked memory. So I was basically operating under the wost-case assumption that each individually mlocked allocation would eat up a whole page of mlocked RAM - this doesn't always happen, but depending on what else is allocating in between, it is pretty likely that most of the things that we need to mlock will occupy a full page.

But I can see how that would be wasteful in a more typical application of a mempool.

In this light, I think we should just keep the code as is, leak the "number of groups" implementation detail, and do the mlock-usecase-specific calculation on the user side, as I have it now.

I don't think it's changing the pool implementation to support arbitrary group sizes is worth it.